sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/NullableKeyKafkaRecordCoder.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kafka;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.beam.sdk.coders.BooleanCoder;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.IterableCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.values.KV;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+
+/** {@link Coder} for {@link KafkaRecord}. */
+@SuppressWarnings({
+  "rawtypes", // TODO(https://issues.apache.org/jira/browse/BEAM-10556)
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class NullableKeyKafkaRecordCoder<K, V> extends StructuredCoder<KafkaRecord<K, V>> {
```

Comment:
Have you considered using the standard NullableCoder ? https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/NullableCoder.java

This will also make it easy to make existing external Kafka transforms support null keys as well (we have to make a NullableCoder a standard coder).

This was discussed in the mailing list as well. https://lists.apache.org/thread.html/r1740cb29e46644a08b2d1755900dd6dcd5c62b937da1a01148edfcb9%40%3Cdev.beam.apache.org%3E

Reply:
+1 to delegate to `NullableCoder`.

Reply:
I didn't try to expose `NullableKeyKafkaRecordCoder` to external usage in this particular pr but I think you made a good point. Thank you! I'll work on that.

Reply:
Thanks.

Reply:
btw I created https://issues.apache.org/jira/browse/BEAM-12585 for tracking `NullableCoder`.

Reply:
Is there a reason to not to do it in this PR ? 
NullableKeyKafkaRecordCoder -> NullableCoder will result in another round of incompatibility, right ?
(I'm fine with NullableKeyKafkaRecordCoder if there's a good reason not to use NullableCoder).

Reply:
(to clarify, making it a standard coder part can be seperate)

Reply:
I believe this PR has used `NullableCoder` and  `NullableKeyKafkaRecordCoder ` should be removed for cleanup. The https://issues.apache.org/jira/browse/BEAM-12585 is for making `NullableCoder` as `StandardCoder`.

Reply:
Ah right. Yeah let's remove NullableKeyKafkaRecordCoder and the test :)